src/PFire.Data/Services/PFireDatabase.cs

```diff
@@ -0,0 +1,225 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.DependencyInjection;
+using PFire.Data.Commands;
+using PFire.Data.Entities;
+using PFire.Data.Models;
+
+namespace PFire.Data.Services
+{
+    public interface IPFireDatabase
+    {
+        Task<UserModel> InsertUser(string username, string password, string salt);
+        Task InsertMutualFriend(UserModel me, UserModel them);
+        Task InsertFriendRequest(UserModel me, UserModel them, string message);
+        Task<UserModel> QueryUser(string username);
+        Task<List<UserModel>> QueryUsers(string username);
+        Task<List<UserModel>> QueryFriends(UserModel user);
+        Task<List<FriendRequestModel>> QueryPendingFriendRequestsSelf(UserModel user);
+        Task<List<FriendRequestModel>> QueryPendingFriendRequests(UserModel otherUser);
+        Task DeletePendingFriendRequest(UserModel me, params FriendRequestModel[] thems);
+        Task UpdateNickname(UserModel user, string nickname);
+    }
+
+    internal class PFireDatabase : IPFireDatabase
```

Comment:
I don't like all this complexity, which I think is coming from trying to abstract away entity framework. If we're going to use EF, I think we should commit to it and just use EF directly, otherwise what's the point?

Inserting a user should be as simple as:
```c#
dbContext.Add(new User {});
await dbContext.SaveChangesAsync()
```

EF is already transaction as the database context uses the unit of work pattern.

Reply:
The way I see it is that the interactions with the database should be _INSANELY_ simple. EF lets us be simple, if we let it and use it to its full capability. 

This is not a project that is going to scale into hundreds of models; the feature set is finite and quite small.

I hope to not come across as flat out rejecting all your hard work because this all must have taken quite a bit of effort.

Reply:
No it's totally fine! That's why this is a draft instead of 'real' PR 🙂  What I did was introduce the EF code I use in my personal projects which I've refined over the years, but if it's too much (and probably is, given there's two tables lol) then it's easier to remove than it is to put in.